source/common/upstream/cluster_manager_impl.cc

```diff
@@ -264,7 +266,27 @@ ClusterManagerStats ClusterManagerImpl::generateStats(Stats::Scope& scope) {
                                     POOL_GAUGE_PREFIX(scope, final_prefix))};
 }
 
-void ClusterManagerImpl::postInitializeCluster(Cluster& cluster) {
+void ClusterManagerImpl::onPerClusterInit(Cluster& cluster) {
+  // This routine is called when a cluster has finished initializing. The cluster has not yet
```

Comment:
Trying to fully grok why the earlier PR had issues with connection/request timeout. A couple of questions:

1. Why did slowness on the main thread (with corresponding libevent cache issues) ever affect the connection/request timeouts, since I would assume these would be a worker thread concern?

2. Wondering what happens when we scale to very large endpoint set sizes. While we no longer block the worker thread to recompute the ring, might we now be spending extended times with an empty host pick (e.g. if we went from no hosts to O(100k) hosts in an update)?

Reply:
>Why did slowness on the main thread (with corresponding libevent cache issues) ever affect the connection/request timeouts, since I would assume these would be a worker thread concern?

The issues are actually on the main thread, not the workers. The main thread is doing a whole bunch of processing around health checking, xDS updates, etc. During initialization, we are very likely to get our xDS responses at the same time (essentially during a single event loop iteration). So the flow looked something like:

1) Initial EDS responses come in.
2) Do membership updates, which used to cause a ring calculation (which takes ~100ms per ring at the sizes we are using on our production boxes x 4 rings on one service = ~400ms). 
3) Setup for HC
4) Then do a timer timeout pass.
5) Oops, all the HC connections timed out.

The way I have it now, during init we only do the ring calculations when a cluster is ready. This basically means that there are no immediate pending init actions that might time out. With that said, I fully admit that this entire situation is generally problematic and may have to become more complicated in the future. In fact, we have seen similar issues already at Lyft where we get a CDS update that adds ~150 clusters, and because of the way stats allocation is occurring creating 150 clusters basically does the same thing and blocks the event loop for too long.

Once the server starts running, events start getting spread out, and spurious timeouts are dealt with gracefully. It's really during the init case that this is a major problem. But again with that said, this is something we are going to have to keep an eye on. Init may need to get more complicated by limiting the amount of work done in each event loop iteration. Futhermore, I think using Ketama at these ring sizes is ultimately not reasonablem, and we should be using a different algorithm (likely Maglev, but need to investigate).

I also spent time thinking about whether to do ring updates on a different thread. I think this is possible, and not that hard to implement. However, there are two reasons this is complicated and I would prefer to think about in a follow up:

1) During the init pass, this makes the logic substantially more complicated as we now have to add async init callbacks for the load balancer also.
2) It makes the subset LB a lot more complicated.

One thing that did occur to me is that we could actually do the initial load synchronously, and then switch to async via worker thread. I will add some TODOs/comments/thoughts around this also.

>Wondering what happens when we scale to very large endpoint set sizes. While we no longer block the worker thread to recompute the ring, might we now be spending extended times with an empty host pick (e.g. if we went from no hosts to O(100k) hosts in an update)?

During initial load, it will be fine, as we don't start listening until everything is initialized. During CDS updates, yes, it may take longer with empty host sets. Though, if the the cluster didn't exist before not sure what the difference is and we can make even smoother by doing CDS warming like we do LDS warming.

Reply:
I would also add that in general, lengthy ring/data structure calculations are bad news if there are a lot of host set updates. I think another thing we need to look into is host set update batching. I don't think this would be hard to implement. E.g., have a setting on a cluster where you say that you batch host set updates for some period of time (say 10ms, 25ms, etc.), and then send out bulk updates. This would help on cluster types where we might see a bunch of churn to prevent cascading updates. This is one area where doing the updates on a worker thread because pending work could be collapsed. I will add all of this to more TODOs/comments.

Reply:
I agree that we can do things on a worker thread. I think outside of server init, it's actually straightforward. I will look at this in a follow up.